File: 지연(이)/24년1월25일/BOJ 테트로미노.py
d = [(0,1),(0,-1),(1,0),(-1,0)]
def dfs(x,y,depth, value):
    global answer
    if depth==3:
        answer = max(answer, value)
        return
    for dx,dy in d:
        nx = x + dx
        ny = y + dy
        if 0<=nx<N and 0<=ny<M and not visit[nx][ny]:
            if depth==1:
                visit[nx][ny]=1
                dfs(x, y, depth+1, value+arr[nx][ny])
                visit[nx][ny]=0
            visit[nx][ny] = 1
            dfs(nx, ny, depth+1, value+arr[nx][ny])
            visit[nx][ny] = 0

import sys
input = sys.stdin.readline
N, M = map(int, input().split())
arr = [list(map(int, input().split())) for _ in range(N)]
visit = [[0]*M for _ in range(N)]
answer = 0
for i in range(N):
    for j in range(M):
        visit[i][j] = 1
        dfs(i,j,0, arr[i][j])
        visit[i][j] = 0

print(answer)

File: 지연(이)/24년1월25일/BOJ 테트로미노.cpp
#include<iostream>
#include<limits.h>

using namespace std;
int N, M;
int arr[501][501];
int visit[501][501];
int answer = INT_MIN;
int dx[4] = {-1,1,0,0};
int dy[4] = {0,0,-1,1};


void DFS(int x, int y, int Sum, int depth) {
	if (depth == 3) {
		answer = max(answer, Sum);
	}
	else {
		for (int k = 0; k < 4; k++) {
			int nx = x + dx[k];
			int ny = y + dy[k];
			if (nx>=0 && nx<N && ny>=0 && ny<M && visit[nx][ny]==0) {
				if (depth == 1) {
					visit[nx][ny] = 1;
					DFS(x, y, Sum + arr[nx][ny], depth + 1);
					visit[nx][ny] = 0;
				}
				visit[nx][ny] = 1;
				DFS(nx, ny, Sum + arr[nx][ny], depth + 1);
				visit[nx][ny] = 0;
			}
		}
	}
	
}

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);


	cin >> N >> M;
	for (int i = 0; i < N; i++) {
		for (int j = 0; j < M; j++) {
			cin >> arr[i][j];
			visit[i][j] = 0;
		}
	}

	for (int i = 0; i < N; i++) {
		for (int j = 0; j < M; j++) {
			visit[i][j] = 1;
			DFS(i, j, arr[i][j], 0);
			visit[i][j] = 0;
		}
	}
	cout << answer << "\n";
	return 0;
}
